src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
-                {
-                    // The remaining 7 or less bits are all 1, which is padding.
-                    // We specifically check that lastDecodedBits > 0 because padding
-                    // longer than 7 bits should be treated as a decoding error.
-                    // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
-                    break;
-                }
-
-                // The longest possible symbol size is 30 bits. If we're at the last 4 bytes
-                // of the input, we need to make sure we pass the correct number of valid bits
-                // left, otherwise the trailing 0s in next may form a valid symbol.
-                int validBits = Math.Min(30, (8 - lastDecodedBits) + (src.Length - i - 1) * 8);
-                int ch = DecodeValue(next, validBits, out int decodedBits);
-
-                if (ch == -1)
-                {
-                    // No valid symbol could be decoded with the bits in next
-                    throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
-                }
-                else if (ch == 256)
-                {
-                    // A Huffman-encoded string literal containing the EOS symbol MUST be treated as a decoding error.
-                    // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
-                    throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
-                }
+                // Load next 8 bits into accumulator.
+                acc <<= 8;
+                acc |= src[i++];
+                bitsInAcc += 8;
 
-                if (j == dst.Length)
+                // Decode bits in accumulator.
+                do
                 {
-                    Array.Resize(ref dstArray, dst.Length * 2);
-                    dst = dstArray;
-                }
+                    lookupIndex = (byte)(acc >> (bitsInAcc - 8));
 
-                dst[j++] = (byte)ch;
+                    ushort lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
 
-                // If we crossed a byte boundary, advance i so we start at the next byte that's not fully decoded.
-                lastDecodedBits += decodedBits;
-                i += lastDecodedBits / 8;
+                    if (lookupValue < 0x80_00)
+                    {
+                        // Octet found.
+                        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                        // | 0 |     number_of_used_bits   |              octet            |
+                        // +---+---------------------------+-------------------------------+
+                        if (j == dst.Length)
+                        {
+                            Array.Resize(ref dstArray, dst.Length * 2);
+                            dst = dstArray;
+                        }
+                        dst[j++] = (byte)lookupValue;
```

Comment:
```suggestion
                        if ((uint)j < (uint)dst.Length)
                        {
                            dst[j++] = (byte)lookupValue;
                        }
                        else
                        {
                            Array.Resize(ref dstArray, dst.Length * 2);
                            dst = dstArray;
                        }
```
This pattern has two points:
* the `uint` casts allow the JIT to eliminate the bound-checks on `dst[j]`
* the first branch is the more likely, and cpu's branch predictor will choose the fallthrough-path when no history for that branch is available

You can use this on other places as well.

Reply:
Ahhh, I've forgotten the write in the case of resizing.
Maybe this cold-portion of code should be used to a non-inlineable method, and there the resive and write after the resize can happen.

The idea is to keep the hot code as small as possible.

Reply:
Why should hot code be as small as possible in this particular case?

Reply:
I have benchmarked recommended changes and did not measured significant perf difference.
I'll keep it as is for code maintainability.,

Reply:
The case `j == dst.Length` is unlikely, at least it will happen less often than the other branch (`j < dst.Length`). In the case resize is needed, we have to pay the cost of resizing anyway, so we could move this part out into a cold region. This will make the hot region smaller, which is a loop (L439-L476) in this case.

Smaller loops means it's more likely that the cpu instructions can be read from an internal buffer / µop-cache or the (instruction) cache(line) instead of farther away memory hierarchies (which are slower) -- in short: the cpu's frontend has less work to do, which improves overall execution speed (also for other parts of the app or other processes). 

Side note: if the loop is small enough (size varies depending on cpu, but say 64 µops) then the cpu has a special loop-stream-detector (LSD). This means the cpu recognizes the loop and is able to use more efficient scheduling of instructions (roughly spoken). But the loops here are too big for this...

Reply:
As hint: in a micro-benchmark the cpu-caches are hot, which means the data and instruction are very likely in the cache (due to the tight benchmark iterations). This may not be the case in real world usage, where the cpu may evict these items from it's caches (by other data and instructions). This also falls into the "general rule" of smaller code is better (increased locality).